pvlib/iotools/bsrn.py

```diff
@@ -0,0 +1,137 @@
+    dhi_std                  float   Std. diffuse horizontal irradiance [W/m^2]
+    dhi_min                  float   Min. diffuse horizontal irradiance [W/m^2]
+    dhi_max                  float   Max. diffuse horizontal irradiance [W/m^2]
+    lwd                      float   Mean. downward long-wave radiation [W/m^2]
+    lwd_std                  float   Std. downward long-wave radiation [W/m^2]
+    lwd_min                  float   Min. downward long-wave radiation [W/m^2]
+    lwd_max                  float   Max. downward long-wave radiation [W/m^2]
+    temp_air                 float   Air temperature [°C]
+    relative_humidity        float   Relative humidity [%]
+    pressure                 float   Atmospheric pressure [hPa]
+    =======================  ======  ==========================================
+
+    References
+    ----------
+    .. [1] `World Radiation Monitoring Center - Baseline Surface Radiation
+        Network (BSRN) <https:/https://bsrn.awi.de/>`_
+    .. [2] `Update of the Technical Plan for BSRN Data Management, 2013,
+       Global Climate Observing System (GCOS) GCOS-172.
+       <https://bsrn.awi.de/fileadmin/user_upload/bsrn.awi.de/Publications/gcos-174.pdf>`_
+    .. [3] `BSRN Data Retrieval via FTP
+       <https://bsrn.awi.de/data/data-retrieval-via-ftp/>`_
+    .. [4] `BSRN Data Release Guidelines
+       <https://bsrn.awi.de/data/conditions-of-data-release/>`_
+    """
+
+    # Read file and store the starting line number for each logical record (LR)
+    line_no_dict = {}
+    if str(filename).endswith('.gz'):  # check if file is a gzipped (.gz) file
+        open_func, mode = gzip.open, 'rt'
+    else:
+        open_func, mode = open, 'r'
+    with open_func(filename, mode) as f:
+        for num, line in enumerate(f):
+            if num == 1:  # Get month and year from the 2nd line
+                start_date = pd.Timestamp(year=int(line[7:11]),
+                                          month=int(line[3:6]), day=1,
+                                          tz='UTC')  # BSRN timestamps are UTC
+            if line.startswith('*'):  # Find start of all logical records
+                line_no_dict[line[2:6]] = num  # key is 4 digit LR number
+
+    # Determine start and end line of logical record LR0100 to be parsed
+    start_row = line_no_dict['0100'] + 1  # Start line number
+    # If LR0100 is the last logical record, then read rest of file
+    if start_row-1 == max(line_no_dict.values()):
+        end_row = num  # then parse rest of the file
+    else:  # otherwise parse until the beginning of the next logical record
+        end_row = min([i for i in line_no_dict.values() if i > start_row])
+    nrows = end_row-start_row
+
+    # Read file as a fixed width file (fwf)
+    data = pd.read_fwf(filename, skiprows=start_row, nrows=nrows, header=None,
+                       colspecs=COL_SPECS, na_values=[-999.0, -99.9])

```

Comment:
It should be possible to avoid reading the file twice by accumulating the first read of just the '0100' section into an IO buffer, stopping the read at the end of the section, and then passing that buffer into `pd.read_fwf`. 

I think this might result in meaningful performance improvements when reading many files. Or maybe we need to profile the code before we put the effort into it.

In any case, I'm ok saving this idea for future work.

Reply:
@wholmgren The initial opening of the file and looping through all of the lines takes ~414 ms, whereas the pd.read_fwf takes 1.15 s. Tested in a Jupyter Notebook using the %%timeit command.

Other parts of the code take on the order of µs (except setting the datetime index 3 ms).

So there are potential savings, however, after searching for a while I still have no idea of how to write to a buffer line by line. I'm also very fine with saving this idea for future work though.

Reply:
Thanks for putting in the effort to profile the code! The potential speed up seems modest so let's save it for future work.